### ISSUE TITLE
performance test between syndtr/goleveldb

### ISSUE THREAD

**Issue by guagualvcha** [cockroachdb/pebble#1167]
Hi, teams. 
Any plan to benchmark pebble db and syndtr/goleveldb?

I tried hard to tuning pebble db, but the performance seems worse than `syndtr/goleveldb`

**Comment by petermattis** [comment]
@guagualvcha We don't have any benchmarks vs `syndtr/goleveldb`, nor do we have any plans to do so. I suspect that you're comparing apples to oranges somehow. For example, perhaps you have `syndtr/goleveldb` to not fsync after each operation. If you can provide a reproducible benchmark showing a performance difference we can possibly spot the cause. As is, this issue isn't actionable.

**Comment by Alex-duzhichao** [comment]
@petermattis @guagualvcha Can you tell me what's the difference and relationship between pebble and goleveldb? I can't find any information on the Internet. Thanks :)

**Comment by petermattis** [comment]
@Alex-duzhichao Pebble and `github.com/syndtr/goleveldb` are both inspired by Google's C++ LevelDB library. Pebble was forked from the incomplete `github.com/golang/leveldb`, but has been essentially completely rewritten. I'm not sure about `syndtr/goleveldb's` heritage.

**Comment by Alex-duzhichao** [comment]
> @Alex-duzhichao Pebble and `github.com/syndtr/goleveldb` are both inspired by Google's C++ LevelDB library. Pebble was forked from the incomplete `github.com/golang/leveldb`, but has been essentially completely rewritten. I'm not sure about `syndtr/goleveldb's` heritage.

@petermattis Clear explanation. Thank you so much.

**Comment by ahothan** [comment]
I have benchmarked pebble, go-leveldb, badger on a very specific use case and found pebble to be the best option of the 3 at very high scale while go-leveldb would perform better at low scale. Badger would be fastest but was using far too much memory as it keeps all keys in memory. So I would say it really depends on the db options as mentioned by @petermattis and on the benchmark conditions.
FWIW, my use case is to have the best check+insert throughput with a very large number of keys (10 billions and more)  where keys are fixed size and values are empty (this is pretty much used as a giant key lookup).

**Comment by zengmenglu** [comment]
> I have benchmarked pebble, go-leveldb, badger on a very specific use case and found pebble to be the best option of the 3 at very high scale while go-leveldb would perform better at low scale. Badger would be fastest but was using far too much memory as it keeps all keys in memory. So I would say it really depends on the db options as mentioned by @petermattis and on the benchmark conditions.
> FWIW, my use case is to have the best check+insert throughput with a very large number of keys (10 billions and more) where keys are fixed size and values are empty (this is pretty much used as a giant key lookup).

Can you tell me what's the ```pebble.Options``` you congfigured in your tests? I'm haunted by the ```pebble.Options``` nowadays. Thanks.

**Comment by ahothan** [comment]
I only tried to run with bloom filters and without, and experimented with 8MB MemTableSize (vs default 4MB). In general impact really varies based on how the db is being used, so what works for me may not work for you.
LSM options is a recurring source of questions because there are lots of options and most users don't have a sufficient understand of LSM to know how to tune them. The general consensus is to try it out on your own workload and see (which can take a lot of time).
I'm currently content with what I have for my use case (no bloom filters and 8MB mem table size) and don't have too much time to experiment more as the results are satisfactory. 
Re. my own benchmark of go-leveldb vs pebble - and this may not be true for other type of workload - I found that go-leveldb was better than pebble on insertion (Add/Set) but worst on a Get. At very high scale (30B keys) I found pebble to be a lot more stable and consistent than go-leveldb for fast storage drives (NVMe) and for slower drives (NetApp)
